tools/auto/presenceValidator.ts

```diff
@@ -0,0 +1,277 @@
+import { createRequire } from "module";
+import { existsSync, readFileSync } from "node:fs";
+import { resolve } from "node:path";
+
+import actions from "@actions/core";
+import got from "got";
+import jsonAst, { ObjectNode } from "json-to-ast";
+import { validate } from "jsonschema";
+import { compare } from "semver";
+
+import PresenceCompiler from "../classes/PresenceCompiler.js";
+import getDiff from "../util/getDiff.js";
+import getLatestSchema from "../util/getLatestSchema.js";
+
+const require = createRequire(import.meta.url),
+	schema = await getLatestSchema(),
+	changedPresences = getDiff(),
+	validLangs = await getValidLanguages(),
+	compiler = new PresenceCompiler(),
+	errors: {
+		presence: string;
+		message: string | Error;
+		properties?: actions.AnnotationProperties | undefined;
+	}[] = [];
+
+for (const presence of changedPresences) {
+	const presencePath = compiler.getPresenceFolder(presence);
+
+	//#region Metadata Check
+	if (!existsSync(resolve(presencePath, "metadata.json"))) {
+		errors.push({
+			presence,
+			message: "Presence is missing metadata.json!",
+			properties: {
+				file: resolve(presencePath, "metadata.json"),
+			},
+		});
+		continue;
+	}
+
+	interface SchemaMetadata extends Metadata {
+		$schema: string;
+	}
+
+	let metadata: SchemaMetadata;
+
+	try {
+		metadata = require(resolve(presencePath, "metadata.json"));
+
+		metadata;
+	} catch {
+		errors.push({
+			presence,
+			message: "Presence metadata.json is not a valid JSON file",
+			properties: {
+				file: resolve(presencePath, "metadata.json"),
+			},
+		});
+		continue;
+	}
+	//#endregion
+
+	const storePresence = (
+		await getStorePresence(metadata.service)
+	).data.presences.find(p => p.metadata.service === metadata.service);
+
+	//#region Schema Check
+	const result = validate(metadata, schema.schema);
+
+	if (result.errors.length) {
+		for (const error of result.errors) {
+			errors.push({
+				presence,
+				message: error.message,
+				properties: {
+					file: resolve(presencePath, "metadata.json"),
+					startLine: getLine(error.argument),
+				},
+			});
+		}
+
+		continue;
+	}
+	//#endregion
+
+	//#region Schema version Check
+	if (metadata.$schema !== schema.url) {
+		errors.push({
+			presence,
+			message: `Schema version is not up to date! - expected: ${schema.url}, got: ${metadata.$schema}`,
```

Comment:
this is not how punctuation works
```suggestion
			message: `Schema version is not up to date - expected: ${schema.url}, got: ${metadata.$schema}`,
```